File: src/main/java/org/modelingvalue/nelumbo/collections/NList.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.collections;

import java.io.Serial;

import org.modelingvalue.collections.List;
import org.modelingvalue.collections.Set;
import org.modelingvalue.nelumbo.AstElement;
import org.modelingvalue.nelumbo.ConstructionReason;
import org.modelingvalue.nelumbo.KnowledgeBase;
import org.modelingvalue.nelumbo.NelumboConstructor;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.NodeInfo;
import org.modelingvalue.nelumbo.lang.Functor;
import org.modelingvalue.nelumbo.lang.Type;
import org.modelingvalue.nelumbo.syntax.ParseContext;
import org.modelingvalue.nelumbo.syntax.ParseException;
import org.modelingvalue.nelumbo.syntax.TokenType;

public final class NList extends Node {
    @Serial
    private static final long serialVersionUID = 2275866157289787141L;

    public NList(List<AstElement> elements, Type elementType) {
        super(NodeInfo.of(elementType.list(), elements), List.of());
    }

    public NList(Type elementType, List<AstElement> elements, List<Node> args) {
        super(NodeInfo.of(elementType.list(), elements), args);
    }

    public NList(Functor functor, List<AstElement> elements, List<Node> args) {
        super(NodeInfo.of(functor, elements), args);
    }

    @NelumboConstructor
    public NList(NodeInfo nodeInfo, Object... args) {
        super(nodeInfo, args);
    }

    public NList(List<AstElement> elements, NList list, Node last) {
        super(NodeInfo.of(list.type(), list.astElements().addAll(elements).add(last)), list.elements().add(last));
    }

    @Override
    protected NList set(NodeInfo nodeInfo, Object[] args) {
        return new NList(nodeInfo, args);
    }

    @Override
    public List<Object> args() {
        return elements();
    }

    @Override
    public NList setAstElements(List<AstElement> elements) {
        return (NList) super.setAstElements(elements);
    }

    public Type elementType() {
        return type().element();
    }

    @SuppressWarnings("unchecked")
    public <T> List<T> elements() {
        return (List<T>) get(0);
    }

    public <T extends Node> List<T> elementsFlattened() {
        List<T> result = List.of();
        for (T e : this.<T>elements()) {
            if (e instanceof NList nl) {
                result = result.addAll(nl.<T>elementsFlattened());
            } else {
                result = result.add(e);
            }
        }
        return result;
    }

    @Override
    public NList set(int i, Object... a) {
        return (NList) super.set(i, a);
    }

    @Override
    public String toString(TokenType[] previous) {
        return elements().toString().substring(4);
    }

    @Override
    public Node add(Node added) {
        return new NList(List.of(), this, added);
    }

    @Override
    public Node init(KnowledgeBase knowledgeBase, ParseContext ctx, ConstructionReason reason) throws ParseException {
        if (reason != ConstructionReason.parsing || (length() > 0 && get(0) instanceof List)) {
            return this;
        }
        return setArgs(new Object[] { Set.of(super.args()) });
    }
}

File: web-ts/src/collections/NList.ts
/**
 * NList - List node type.
 * @JAVA_REF org.modelingvalue.nelumbo.collections.NList
 */

import { List } from 'immutable';
import type { AstElement } from '../AstElement';
import { Type } from '../Type';
import { Node } from '../Node';
import type { Functor } from '../patterns/Functor';
import { TokenType } from '../syntax/TokenType';

export class NList extends Node {
  constructor(elements: List<AstElement>, elementType: Type);
  constructor(elementType: Type, elements: List<AstElement>, args: List<Node>);
  constructor(functor: Functor, elements: List<AstElement>, args: unknown[]);
  constructor(elements: List<AstElement>, list: NList, last: Node);
  constructor(elements: List<AstElement>, elementType: Type, ...nodes: Node[]);
  constructor(
    arg1: List<AstElement> | Type | Functor,
    arg2: Type | List<AstElement> | NList,
    arg3?: List<Node> | unknown[] | Node,
    ...rest: Node[]
  ) {
    if (arg1 instanceof Type) {
      // NList(elementType, elements, args)
      const elementType = arg1;
      const elements = arg2 as List<AstElement>;
      const args = arg3 as List<Node>;
      super(elementType.list(), elements, args);
    } else if (arg2 instanceof Type && (arg3 !== undefined || rest.length > 0)) {
      // NList(elements, elementType, ...nodes)
      const elements = arg1 as List<AstElement>;
      const elementType = arg2;
      const nodes: Node[] = [];
      if (arg3 !== undefined) {
        nodes.push(arg3 as Node);
      }
      nodes.push(...rest);
      super(elementType.list(), elements, List(nodes));
    } else if (arg2 instanceof Type) {
      // NList(elements, elementType) - no additional args
      const elements = arg1 as List<AstElement>;
      const elementType = arg2;
      super(elementType.list(), elements, List<Node>());
    } else if (arg2 instanceof NList) {
      // NList(elements, list, last)
      const elements = arg1 as List<AstElement>;
      const list = arg2;
      const last = arg3 as Node;
      super(list.type(), list.astElements().concat(elements).push(last as AstElement), list.elements().push(last));
    } else if ('pattern' in arg1) {
      // NList(functor, elements, args)
      const functor = arg1 as Functor;
      const elements = arg2 as List<AstElement>;
      const args = arg3 as unknown[];
      super(functor, elements, List(args));
    } else {
      // Fallback
      const elements = arg1 as List<AstElement>;
      super(Type.ROOT.list(), elements, List<Node>());
    }
  }

  protected static fromData(data: unknown[], declaration?: Node): NList {
    const node = Object.create(NList.prototype) as NList;
    (node as unknown as { _data: unknown[] })._data = data;
    (node as unknown as { _declaration: Node | undefined })._declaration = declaration ?? node;
    return node;
  }

  protected override struct(data: unknown[], declaration?: Node): Node {
    return NList.fromData(data, declaration ?? this.declaration());
  }

  override args(): List<unknown> {
    return this.elements();
  }

  override setAstElements(elements: List<AstElement>): NList {
    return super.setAstElements(elements) as NList;
  }

  elementType(): Type {
    return this.type().element()!;
  }

  elements<T = Node>(): List<T> {
    return this.get(0) as List<T>;
  }

  elementsFlattened<T extends Node = Node>(): List<T> {
    let result = List<T>();
    for (const e of this.elements<T>()) {
      if (e instanceof NList) {
        result = result.concat(e.elementsFlattened<T>());
      } else {
        result = result.push(e);
      }
    }
    return result;
  }

  override toString(_previous?: TokenType[]): string {
    const str = this.elements().toString();
    // Remove "List" prefix from Immutable.js toString
    return str.startsWith('List') ? str.substring(4) : str;
  }
}
